File: tests/cli/integration/hook-byte-equality.test.ts
/**
 * E6 fitness function: the installed pre-commit hook must be byte-identical
 * to the verbatim R25a script across releases.
 *
 * The integration test spawns the built CLI, runs `hook install`, and
 * asserts the on-disk file matches the canonical bytes. This is the
 * acceptance gate that catches a release-time spec edit (a CI job calls
 * this and fails the build if the bytes drift without a deliberate bump).
 */

import { promises as fs } from "node:fs";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { HOOK_SCRIPT } from "../../../src/cli/commands/hook.js";
import { makeFixture, type Fixture } from "../../helpers/fixture.js";

import { ensureBuilt, runCli } from "./spawn.js";

let fx: Fixture | undefined;
afterEach(async () => {
  if (fx) await fx.cleanup();
  fx = undefined;
});

describe("R25a hook script byte equality (E6 fitness function)", () => {
  it("hook install writes the verbatim spec script", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    await fs.mkdir(path.join(fx.projectRoot, ".git", "hooks"), { recursive: true });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "hook", "install"],
    });
    expect(r.exitCode).toBe(0);
    const written = await fs.readFile(
      path.join(fx.projectRoot, ".git", "hooks", "pre-commit"),
      "utf8",
    );
    expect(written).toBe(HOOK_SCRIPT);
    // Pin against the literal string so a typo in HOOK_SCRIPT itself can't
    // pass the round-trip silently (the fitness function asserts both the
    // constant *and* the on-disk content match the spec).
    expect(written).toBe(
      "#!/bin/sh\n" +
        "command -v c3p >/dev/null 2>&1 || exit 0\n" +
        "c3p drift --pre-commit-warn 2>&1\n" +
        "exit 0\n",
    );
  });

  it("init writes the same hook bytes via the bootstrap path", async () => {
    await ensureBuilt();
    fx = await makeFixture({});
    await fs.mkdir(path.join(fx.projectRoot, ".git", "hooks"), { recursive: true });
    const r = await runCli({
      args: ["--cwd", fx.projectRoot, "init", "--no-seed"],
    });
    expect(r.exitCode).toBe(0);
    const written = await fs.readFile(
      path.join(fx.projectRoot, ".git", "hooks", "pre-commit"),
      "utf8",
    );
    expect(written).toBe(HOOK_SCRIPT);
  });
});

File: go/tests/integration/hook_byte_equality_test.go
// hook_byte_equality_test.go — narrow exemption to the spawn-only contract
// documented in scenarios_test.go. The byte-equality tests below are E6
// fitness functions that pin the EXACT hook-script bytes against a known
// constant; importing the constants from internal/cli/commands is the only
// way to spell "the bytes the bin would write" without copying them (which
// would defeat the regression purpose). This is the only test file allowed
// to import internal product packages — every other test in this directory
// must drive the bin via spawn.
package integration_test

import (
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/htxryan/c3p/internal/cli/commands"
	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestHookByteEquality_HookInstallWritesCanonicalBytes — IV translation
// of TS hook-byte-equality.test.ts. The pre-commit hook installed via
// `c3p hook install` must be byte-identical to commands.HookScriptContent
// (R25a + Go's PR15 spec). This is an E6 fitness function: a release-time
// edit to the script bytes is a deliberate spec bump, not a free change.
func TestHookByteEquality_HookInstallWritesCanonicalBytes(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	if err := os.MkdirAll(filepath.Join(fx.ProjectRoot, ".git", "hooks"), 0o755); err != nil {
		t.Fatalf("mkdir .git/hooks: %v", err)
	}
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "hook", "install"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("hook install: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	written, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".git", "hooks", "pre-commit"))
	if err != nil {
		t.Fatalf("read pre-commit: %v", err)
	}
	if string(written) != commands.HookScriptContent {
		t.Fatalf("pre-commit content drift:\nwant: %q\ngot:  %q",
			commands.HookScriptContent, string(written))
	}
}

// TestHookByteEquality_InitWritesSameBytes — init's bootstrap path
// (`init --no-seed`) installs the hook by default; the bytes must match
// the standalone `hook install` path.
func TestHookByteEquality_InitWritesSameBytes(t *testing.T) {
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	if err := os.MkdirAll(filepath.Join(fx.ProjectRoot, ".git", "hooks"), 0o755); err != nil {
		t.Fatalf("mkdir .git/hooks: %v", err)
	}
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "init", "--no-seed"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("init: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	written, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".git", "hooks", "pre-commit"))
	if err != nil {
		t.Fatalf("read pre-commit: %v", err)
	}
	if string(written) != commands.HookScriptContent {
		t.Fatalf("init pre-commit drift:\nwant: %q\ngot:  %q",
			commands.HookScriptContent, string(written))
	}
}

// TestHookByteEquality_BatCompanionOnWindows — PR15: on Windows, both
// pre-commit and pre-commit.bat are written, with .bat byte-identical
// to commands.HookScriptContentBat. CRLF intentional (cmd.exe parser).
func TestHookByteEquality_BatCompanionOnWindows(t *testing.T) {
	if runtime.GOOS != "windows" {
		t.Skip("PR15 .bat companion is Windows-only")
	}
	helpers.EnsureBuilt(t)
	fx := helpers.MakeFixture(t, helpers.FixtureSpec{})
	if err := os.MkdirAll(filepath.Join(fx.ProjectRoot, ".git", "hooks"), 0o755); err != nil {
		t.Fatalf("mkdir .git/hooks: %v", err)
	}
	r := mustRun(t, helpers.SpawnOptions{
		Args: []string{"--cwd", fx.ProjectRoot, "hook", "install"},
	})
	if r.ExitCode != 0 {
		t.Fatalf("hook install: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	written, err := os.ReadFile(filepath.Join(fx.ProjectRoot, ".git", "hooks", "pre-commit.bat"))
	if err != nil {
		t.Fatalf("read pre-commit.bat: %v", err)
	}
	if string(written) != commands.HookScriptContentBat {
		t.Fatalf("pre-commit.bat content drift:\nwant: %q\ngot:  %q",
			commands.HookScriptContentBat, string(written))
	}
}
